services/galley/src/Galley/API/Update.hs

```diff
@@ -333,73 +326,71 @@ joinConversation zusr zcon cnv access = do
     ensureAccessRole (Data.convAccessRole conv) [zusr] mbTms
     let newUsers = filter (notIsMember conv) [zusr]
     ensureMemberLimit (toList $ Data.convMembers conv) newUsers
-    addToConversation (botsAndUsers (Data.convMembers conv)) zusr zcon newUsers conv
+    -- NOTE: When joining conversations, all users become members
+    addToConversation (botsAndUsers (Data.convMembers conv)) zusr zcon newUsers conv roleNameWireMember
 
 addMembers :: UserId ::: ConnId ::: ConvId ::: JsonRequest Invite -> Galley Response
 addMembers (zusr ::: zcon ::: cid ::: req) = do
     body <- fromJsonBody req
     conv <- Data.conversation cid >>= ifNothing convNotFound
     let mems = botsAndUsers (Data.convMembers conv)
+    ensureActionAllowed AddConversationMember =<< getSelfMember zusr (snd mems)
     toAdd <- fromMemberSize <$> checkedMemberAddSize (toList $ invUsers body)
     let newUsers = filter (notIsMember conv) (toList toAdd)
     ensureMemberLimit (toList $ Data.convMembers conv) newUsers
     ensureAccess conv InviteAccess
     case Data.convTeam conv of
         Nothing -> do
-            ensureConvMember (snd mems) zusr
             ensureAccessRole (Data.convAccessRole conv) newUsers Nothing
             ensureConnected zusr newUsers
         Just ti -> teamConvChecks ti newUsers conv
-    addToConversation mems zusr zcon newUsers conv
+    addToConversation mems zusr zcon newUsers conv (invRoleName body)
```

Comment:
This is insufficient: invRoleName should not be blindly trusted: 

* If this is a non-team conversation, ignore this and hardcode to wireAdmin. (don't kill democracy! :smile: )
* If this is a team conversation, first check this role exists, and is valid, and doesn't contain higher permission sets than the user granting, before proceeding
* If this is unset, set the default (wireAdmin for non-team conversations, wireMember for team conversations?), clearly explained with a comment.

Reply:
The check for elevation was added with `ensureConvRoleNotElevated`, thanks for pointing that out!